File: bot/cogs/core/voice.py
# ---------------------------------- Imports --------------------------------- #
import re

import discord
from discord.ext import commands
from discord_slash import SlashContext
import lavalink
from lavalink import DefaultPlayer

from ... import config

# ---------------------------------- Config ---------------------------------- #
cfg = config.load_config()
standard_rx = re.compile(r'https?://(?:www\.)?.+')

# --------------------------------- Cog class -------------------------------- #
class VoiceStateManager(commands.Cog):
    def __init__(self, bot):
        self.bot: discord.Client = bot

    # ---------------------------------------------------------------------------- #
    #                              Methods / Commands                              #
    # ---------------------------------------------------------------------------- #

    # ---------------------------- ANCHOR UPDATE STATUS -------------------------- #
    # Updates boomers status message with whatever song is currently playing

    async def update_status(self, player: DefaultPlayer):

        if player.fetch("repeat_one"):
            suffix = " (on repeat)"
        elif player.shuffle:
            suffix = " (shuffle)"
        else:
            suffix = ""

        if player.current:
            activity = discord.Game(player.current.title + suffix)
            status = discord.Status.online

        else:
            activity = discord.Game("nothing.")
            status = discord.Status.idle
            
        await self.bot.change_presence(activity=activity, status=status)

    # ---------------------------- ANCHOR FETCHPLAYER ---------------------------- #
    # Returns the bots player if one exists otherwise sets up the default one and
    # returns that

    def fetch_player(self, ctx: SlashContext) -> lavalink.DefaultPlayer:
        # Create returns a player if one exists, otherwise creates.
        return self.bot.lavalink.player_manager.create(
            ctx.guild.id, endpoint=str(ctx.guild.region)
        )

    # ---------------------------- ANCHOR ENSUREVOICE ---------------------------- #
    # Makes sure boomer joins a voice channel if commands that need it are queued (
    # i.e. anything that plays or queues a song). Otherwise makes sure you're in the 
    # the same voice channel as the bot to run things.

    async def ensure_voice(self, ctx: SlashContext) -> lavalink.DefaultPlayer:
        # This check ensures that the bot and command author are in the same voicechannel.

        player = self.fetch_player(ctx)

        # These are commands that require the bot to join a voicechannel.
        should_connect = ctx.command in ('play','ok','lofi','test','best','load','die')

        if not ctx.author.voice or not ctx.author.voice.channel:
            await ctx.send('Join a voicechannel first.')
            return

        if not player.is_connected:
            if not should_connect:
                await ctx.send('Not connected.')
                return

            player.store('pages', 0)
            player.store('channel', ctx.channel.id)
            player.store('voice', ctx.author.voice.channel)
            await ctx.guild.change_voice_state(channel=ctx.author.voice.channel)

            return player
        else:
            if int(player.channel_id) != ctx.author.voice.channel.id:
                await ctx.send(f"You need to be in <#{player.channel_id}> to do that.")
                return

            return player
        
    # ----------------------------- ANCHOR DISCONNECT ---------------------------- #
    # Safely disconnects the bot from it's voice channel and clears up / resets any 
    # relevent settings.

    async def disconnect(self, player: lavalink.DefaultPlayer, guild: discord.Guild):

        player.queue.clear()
        await player.stop()
        await guild.change_voice_state(channel=None)
        player.store('repeat_one', None)
        player.set_repeat(False)
        player.set_shuffle(False)
        await player.set_volume(cfg['music']['volume_default'])
        await self.update_status(player)
    
VSM = VoiceStateManager

def setup(bot):
    bot.add_cog(VoiceStateManager(bot))

File: src/util/helpers/voice.ts
import { ExtendedClient } from "../structures"
import { CommandInteraction, GuildMember } from "discord.js"
import { Player } from "erela.js"

import config from "../../config.json"

/**
 * Helps manage the voice state with common guards and convenient connection/disconnection methods 
 */
export class VoiceHelper {

    client: ExtendedClient

    /**
     * Note: Any command using ensureVoice is deferred to account for lavalink response times.
     * use editReply for any ineraction reply after ensureVoice runs.
     * @param client modified discord.js client called 'Boomer'.
     */
    constructor(client: ExtendedClient) {
        this.client = client
    }

    /**
     * make sure you never run this in a scenario where a player doesnt exist.
     * inbuilt guard for that coming soon.
     * @param client Custom discord.js client Boomer
     * @returns erela.js player
     */
    static fetchPlayer(client: ExtendedClient) {
        return client.manager.get(config.bot.guildID) as Player
    }

    /**
     * multi purpose catch-all meant to go before a command. 
     * checks that the interaction is within a guild and inside
     * a voice channel. fetches existing player if there is 
     * one else creates a new one using details from given interaction.
     * @param interaction Discord.js CommandInteraction
     * @returns erela.js player
     */
    async ensureVoice(interaction: CommandInteraction) {

        // sender validation: in guild
        if (!(interaction.member instanceof GuildMember) || !interaction.guild || !interaction.channel) {
            interaction.reply({ content: "Try sending this from within a server.", ephemeral: true });
            return
        }

        // sender validation: in voice
        if (!interaction.member.voice.channel) {
            interaction.reply({ content: "You need to join a voice channel.", ephemeral: true });
            return
        }

        // this fetches the player if one exists. otherwise gerenates a new one
        const player = this.client.manager.create({
            guild: interaction.guild.id,
            voiceChannel: interaction.member.voice.channel.id,
            textChannel: interaction.channel.id,
        });

        if (player.state == "DISCONNECTED") {
            // set up some default states and connect the player to requesters VC
            player.set("pages", 0)
            player.set("idle", false)
            player.connect()
            
        } else if (player.voiceChannel !== interaction.member.voice.channel.id) {
            // ensure player is connected to the same VC as the member interacting with it
            await interaction.reply({ content: `Already in <#${player.voiceChannel}> :rolling_eyes:`, ephemeral: true})
            await interaction.followUp({content: `${config.bot.name} can't be in two places at once. Join the linked channel to use ${config.bot.pronoun}.`, ephemeral: true})

        }

        // success! this should be the output at all times :)))
        return player
    }

    /**
     * Resets the state of the player back to its defaults.
     * @param player erela.js player
     */
    async disconnect(player: Player) {
        player.queue.clear()
        player.stop()
        player.setTrackRepeat(false)
        player.setQueueRepeat(false)
        player.setVolume(config.music.volumeDefault)
        player.disconnect()
        player.destroy()
        await VoiceHelper.updateStatus(this.client, player)
        return
    }

    /**
     * Updates the activity and status to match what boomer is doing.
     * playing music: online & shows 'listening to (song). idle: shows 
     * nothing and idle status.
     * @param player erela.js player
     */
    static async updateStatus(client: ExtendedClient, player: Player) {

        var suffix = ""

        // attached modidifiers present to the activity message
        if (player.trackRepeat) {
            suffix = " (on repeat)"
        }

        if (player.playing) {
            // update status based on current song

            client.user!.setActivity(
                player.queue.current!.title + suffix,
                {
                    type: 'LISTENING'
                }
            )
            client.user!.setStatus("online")
            
        } else {
            // update status to idle and show listening to 'nothing'

            client.user!.setActivity(
                "nothing.",
                {
                    type: "LISTENING"
                }
            )
            client.user!.setStatus("idle")
        }

        console.log(`Updated activity info to: ${client.user!.presence.activities}`);
        console.log(`Updated status info to: ${client.user!.presence.status}`)
        return
    }
}